src/role_scout.js

```diff
@@ -13,153 +13,219 @@ roles.scout.settings = {
+  };
 
-    if (creep.isStuck()) {
-      creep.moveRandom();
-      creep.say('ImStuck', true);
-      creep.log('Scout Stuck, Randomly Moving: ' + JSON.stringify(creep.memory.last) + ' ' + JSON.stringify(creep.isStuck()));
-      return true;
+  let search = PathFinder.search(
+    creep.pos, {
+      pos: targetPos,
+      range: 20
+    }, {
+      roomCallback: costMatrix
     }
+  );
+
+  if (config.visualizer.enabled && config.visualizer.showPathSearches) {
+    visualizer.showSearch(search);
+  }
 
-    if (search.path.length === 0 || (creep.inBase() && creep.room.memory.misplacedSpawn)) {
-      creep.say('hello', true);
-      //       creep.log(creep.pos + ' ' + targetPosObject + ' ' + JSON.stringify(search));
-      if (creep.isStuck() && onBorder(creep)) {
-        creep.say('imstuck at the border', true);
-        if (config.room.scoutSkipWhenStuck) {
-          creep.say('skipping', true);
-          creep.memory.scoutSkip = true;
-          delete creep.memory.last; // Delete to reset stuckness.
-        }
+  if (search.path.length === 0 || (creep.inBase() && creep.room.memory.misplacedSpawn)) {
+    if (creep.isStuck() && creep.pos.isBorder()) {
+      creep.say('imstuck at the border', true);
+      if (config.scout.scoutSkipWhenStuck) {
+        creep.say('skipping', true);
+        creep.memory.scoutSkip = true;
+        delete creep.memory.last; // Delete to reset stuckness.
       }
-      //if (search.path.length > 0) {
-      //creep.move(creep.pos.getDirectionTo(search.path[0]));
-      //} else {
-      let returnCode = creep.moveTo(targetPosObject, {
-        ignoreCreeps: true,
-        costCallback: creep.room.getCostMatrixCallback()
-      });
-      //}
-      return true;
     }
-    creep.say(creep.pos.getDirectionTo(search.path[0]));
-    let returnCode = creep.move(creep.pos.getDirectionTo(search.path[0]));
-  };
 
-  creep.notifyWhenAttacked(false);
-  return breadthFirstSearch(creep);
+    let returnCode = creep.moveTo(targetPos, {
+      ignoreCreeps: true,
+      costCallback: creep.room.getCostMatrixCallback()
+    });
+
+    return true;
+  }
+  creep.say('Hi -> ' + creep.memory.search.target, true);
+  let returnCode = creep.move(creep.pos.getDirectionTo(search.path[0]));
+};
+
+roles.scout.breadthFirstSearch = function(creep) {
+  roles.scout.initSearch(creep);
+
+  if (creep.memory.scoutSkip || creep.room.name === creep.memory.search.target) {
+    roles.scout.switchTarget(creep);
+  }
+
+  if (!creep.memory.search.target) {
+    creep.log('Suiciding: ' + JSON.stringify(creep.memory.search));
+    creep.suicide();
+    return true;
+  }
+
+  let targetPos = new RoomPosition(25, 25, creep.memory.search.target);
+
+  if (creep.room.getEnemys().length) {
+    creep.say('Afraid', true);
+    creep.memory.scoutSkip = true;
+    return false;
+  }
+  if (creep.memory.last && creep.memory.last.pos3 && creep.room.name !== creep.memory.last.pos3.roomName) {
+    let roomMem = creep.room.memory;
```

Comment:
just to be clear, creep.room.memory may also be undefined if we see this room for first time. Not sure if you initialize it somewhere else)

Reply:
yeah... will verify ^^'

Reply:
as rooms are handle before creeps in, the memory is already filled with state and lastSeen at least